htmd/parameterization/esp.py

```diff
@@ -103,6 +191,170 @@ def _generate_points(molecule, vdw_radii=[1.4, 1.6, 1.8, 2.0, 2.2], density=10):
 
         return points
 
+    def __init__(self):
+
+        self.molecule = None
+        self.qm_results = None
+        self.fixed = []
+
+        self._reciprocal_distances = None
+
+    @property
+    def ngroups(self):
+        """Number of charge groups"""
+
+        return len(self.molecule._equivalent_atom_groups)
+
+    def _map_groups_to_atoms(self, group_charges):
+
+        charges = np.zeros(self.molecule.natoms)
+        for atom_group, group_charge in zip(self.molecule._equivalent_atom_groups, group_charges):
+            charges[atom_group] = group_charge
+
+        return charges
+
+    def _compute_constraint(self, group_charges, _):
+
+        charges = self._map_groups_to_atoms(group_charges)
+        constraint = np.sum(charges) - self.molecule.netcharge
+
+        return constraint
+
+    def _get_bounds(self):
+
+        # Set bound arrays
+        lower_bounds = np.ones(self.ngroups) * -1.25
+        upper_bounds = np.ones(self.ngroups) * +1.25
+
+        # If the restraint relates to an H, set the lower bound to 0
+        for i in range(self.ngroups):
+            if 'H' == self.molecule.element[self.molecule._equivalent_atom_groups[i][0]]:
+                lower_bounds[i] = 0.001
+
+        # Fix the charges of the specified atoms to those already set in the
+        # charge array. Note this also fixes the charges of the atoms in the
+        # same equivalency group.
+        for atom in self.fixed:
+            group = self.molecule._equivalent_group_by_atom[atom]
+            lower_bounds[group] = self.molecule.charge[atom]
+            upper_bounds[group] = self.molecule.charge[atom]
+
+        return lower_bounds, upper_bounds
+
+    def _compute_objective(self, group_charges, _):
```

Comment:
Same here. You really have to explain me the purpose of this parameter that is not used in the function

Reply:
This is a convention of *NLopt* (https://nlopt.readthedocs.io/en/latest/NLopt_Python_Reference/#Objective_function). It still pass `None` for gradient, even if it is not used.